Bluetooth: Controller: Vendor Specific: Scan Request Reports for Legacy advertising

giansta opened the pull request:
Add debug scan callback in legacy advertisements, meant for
manufacturing or debugging usage. Enabled via proper vendor-specific
command and via debug API, allows saving 5.5 kB RAM that would be taken
when using extended advertisements.

sjanc commented:
How do you enable scan req notifications if ext adv is disabled?

Also, spec is quite clear that controller shall not send LE Scan Request Received event when legacy advertising HCI commands are used:

7.7.65.19 LE Scan Request Received event
"This event shall only be generated if advertising was enabled using the HCI_LE_Set_Extended_Advertising_Enable command."

cvinayak reviewed (changes requested):
Change in the PR are not Bluetooth Core Specification complaint. Suggest introducing vendor specific commands instead, and design implementation changes as necessary to enable  and generate vendor specific scan request reports.

giansta commented:
> Change in the PR are not Bluetooth Core Specification complaint. Suggest introducing vendor specific commands instead, and design implementation changes as necessary to enable and generate vendor specific scan request reports.

Thanks, @sjanc and @cvinayak, for your observations and sorry for my delay.
I tried such a change because our application using nRF52832 implements a BLE central that already uses most of the available 64k SRAM. 
We're optimizing memory usage everywhere, even in the latest added feature that implements an additional broadcaster role to be used for field operations. For this role we advertise with low power (using BT_CTLR_TX_PWR_DYNAMIC_CONTROL) to allow only close scanners to send a scan request and we need the corresponding callback specifically for field operations. 
So your suggestion of using a vendor-specific debug command seems right for our purpose and hopefully it seems to keep Bluetooth complaint for normal operations.

giansta commented:
> @giansta I see the PR changes are moving in the right direction, could you add a simple sample or test covering the enhancement. Also, I think the documentation for VS commands also need update.

Would it be good a sample app folder similar to `samples/bluetooth/hci_pwr_ctrl/`, e.g. something like `samples/bluetooth/hci_vs_scan_req/` (whatever naming suggestion is welcome)?
Yet, to limit time consumption, I would test only it on my current board.
About documentation of VS commands, I couldn't find it, could you please point me to it?

cvinayak commented on include/zephyr/bluetooth/hci_vs.h:
```diff
@@ -194,23 +194,29 @@ struct bt_hci_rp_vs_read_usb_transport_mode {
 	uint8_t  min_used_chans;
 } __packed;
 
+#define BT_HCI_OP_VS_SET_SCAN_REQ_RCVD_MASK     BT_OP(BT_OGF_VS, 0x0013)
+
+struct bt_hci_cp_vs_set_scan_req_rcvd_mask {
+	uint8_t  enable;
+} __packed;
+
```
isnt this already here and documented here: https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/connectivity/bluetooth/api/hci.txt#L763-L803

giansta replied:
Thank you!

cvinayak commented:
> > @giansta I see the PR changes are moving in the right direction, could you add a simple sample or test covering the enhancement. Also, I think the documentation for VS commands also need update.
> 
> Would it be good a sample app folder similar to `samples/bluetooth/hci_pwr_ctrl/`, e.g. something like `samples/bluetooth/hci_vs_scan_req/` (whatever naming suggestion is welcome)? Yet, to limit time consumption, I would test only it on my current board. About documentation of VS commands, I couldn't find it, could you please point me to it?

i see that scan_req is a feature already documented... you should use the already documented opcode.

The documentation is here: https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/connectivity/bluetooth/api/hci.txt#L763-L803

Regarding the sample, i say you start with one PR with only HCI/Controller changes and a sample to use HCI VS commands, get the events. A separate PR for Host stack API introduction.

for the Controller only change, `samples/bluetooth/hci_vs_scan_req/`  sounds good.

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2534,6 +2534,12 @@ static const struct event_handler meta_events[] = {
 #endif /* CONFIG_BT_CONN */
 #endif /* defined(CONFIG_BT_PER_ADV_SYNC) */
 #endif /* defined(CONFIG_BT_EXT_ADV) */
+#if defined(CONFIG_BT_CTLR_SCAN_REQ_NOTIFY)
+#if defined(CONFIG_BT_BROADCASTER)
+	EVENT_HANDLER(BT_HCI_EVT_LE_SCAN_REQ_RECEIVED, bt_hci_le_scan_req_received,
+		      sizeof(struct bt_hci_evt_le_scan_req_received)),
+#endif
+#endif /* defined(CONFIG_BT_CTLR_SCAN_REQ_NOTIFY) */
```
There is already a EVENT_HANDLER for BT_HCI_EVT_LE_SCAN_REQ_RECEIVED upstream, the code here seems redundant.

giansta replied:
Thanks for spotting this, it messed up with my different versions and missed to remove this. Fixed, working to the rest.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -340,7 +340,7 @@ uint8_t ll_adv_params_set(uint16_t interval, uint8_t adv_type,
 	adv->lll.filter_policy = filter_policy;
 
-#if defined(CONFIG_BT_CTLR_SCAN_REQ_NOTIFY)
+#if defined(CONFIG_BT_CTLR_SCAN_REQ_NOTIFY) && defined(CONFIG_BT_CTLR_ADV_EXT)
```
Why `CONFIG_BT_CTLR_ADV_EXT` ?

giansta replied:
It's because I'm setting `CONFIG_BT_CTLR_SCAN_REQ_NOTIFY=y` independently of `CONFIG_BT_CTLR_ADV_EXT`, as it seems possible. (Before I did it in the application, now it's selected by `BT_CTLR_VS_SCAN_REQ_RX`).

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4951,6 +4951,20 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
 #endif /* CONFIG_BT_CTLR_TX_PWR_DYNAMIC_CONTROL */
 
+#if defined(CONFIG_BT_CTLR_SCAN_REQ_NOTIFY)
+static void vs_set_scan_req_rcvd_mask(struct net_buf *buf, struct net_buf **evt)
```
```suggestion
static void vs_set_scan_req_reports(struct net_buf *buf, struct net_buf **evt)
```

giansta replied:
Done, thanks.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -340,7 +340,7 @@ uint8_t ll_adv_params_set(uint16_t interval, uint8_t adv_type,
 
-#if defined(CONFIG_BT_CTLR_SCAN_REQ_NOTIFY)
+#if defined(CONFIG_BT_CTLR_SCAN_REQ_NOTIFY) && defined(CONFIG_BT_CTLR_ADV_EXT)
 	adv->lll.scan_req_notify = sreq;
```
Override using a variable that be set in `vs_set_scan_req_reports`  when it is called.

giansta replied:
If it's legit setting `CONFIG_BT_CTLR_SCAN_REQ_NOTIFY=y` independently of `CONFIG_BT_CTLR_ADV_EXT`, as noted above, I would leave as it is.
When `vs_set_scan_req_reports is called`, I'm using corresponding bit in `vs_events_mask`, as said below.

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4951,6 +4951,20 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+#if defined(CONFIG_BT_CTLR_SCAN_REQ_NOTIFY)
+static void vs_set_scan_req_rcvd_mask(struct net_buf *buf, struct net_buf **evt)
+{
+	struct bt_hci_cp_vs_set_scan_req_rcvd_mask *cmd = (void *)buf->data;
```
```suggestion
	struct bt_hci_cp_vs_set_scan_req_reports *cmd = (void *)buf->data;
```

giansta replied:
Done, thanks.

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4951,6 +4951,20 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+{
+	struct bt_hci_cp_vs_set_scan_req_rcvd_mask *cmd = (void *)buf->data;
+
+	if (cmd->enable != 0) {
+		le_event_mask |= BT_EVT_MASK_LE_SCAN_REQ_RECEIVED;
+	} else {
+		le_event_mask &= ~BT_EVT_MASK_LE_SCAN_REQ_RECEIVED;
+	}
```
Set a variable that can be used to override `adv->lll.scan_req_notify` assignment.

giansta replied:
I've used the corresponding bit in `vs_events_mask`.

cvinayak commented on include/zephyr/bluetooth/bluetooth.h:
This should be separate PR for host stack changes, if at all required.

giansta replied:
I can do that, including the sample app in the 2nd PR, as said above.

cvinayak commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -952,6 +952,18 @@ config BT_CTLR_ASSERT_HANDLER
 	  and will be invoked whenever the controller code encounters
 	  an unrecoverable error.
 
+if !BT_EXT_ADV
+
+config BT_CTLR_VS_SCAN_REQ_RX
+	bool "Use scan request reporting"
+	depends on BT_HCI_VS_EXT
+	select BT_HCI_VS_EVT
+	select BT_CTLR_SCAN_REQ_NOTIFY
+	help
+	  Enables usage of VS Scan Request Reports Command and Scan Request Received Event
+
+endif # !BT_EXT_ADV
+
```
```suggestion
config BT_CTLR_VS_SCAN_REQ_RX
	bool "Use scan request reporting"
	depends on BT_HCI_VS_EXT && !BT_CTLR_ADV_EXT
	select BT_CTLR_SCAN_REQ_NOTIFY
	help
	  Enables usage of VS Scan Request Reports Command and Scan Request Received Event

```

giansta replied:
Thanks, done

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -7951,6 +7975,7 @@ static void le_big_terminate(struct pdu_data *pdu,
 #endif /* CONFIG_BT_BROADCASTER */
 
 #if defined(CONFIG_BT_CTLR_SCAN_REQ_NOTIFY)
+#if defined(CONFIG_BT_EXT_ADV)
```
```suggestion
#if defined(CONFIG_BT_CTLR_ADV_EXT)
```

giansta replied:
Thanks, done

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -8004,6 +8029,50 @@ static void le_scan_req_received(struct pdu_data *pdu_data,
 		       sizeof(bt_addr_t));
 	}
 }
+#endif /* CONFIG_BT_EXT_ADV */
+
```
```suggestion
#endif /* CONFIG_BT_CTLR_ADV_EXT */

```

giansta replied:
Thanks, done

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -8380,7 +8449,13 @@ static void encode_control(struct node_rx_pdu *node_rx,
 
 #if defined(CONFIG_BT_CTLR_SCAN_REQ_NOTIFY)
 	case NODE_RX_TYPE_SCAN_REQ:
+#if defined(CONFIG_BT_EXT_ADV)
```
```suggestion
#if defined(CONFIG_BT_CTLR_ADV_EXT)
```

giansta replied:
Thanks, done. However I cannot find where are the tests with BSIM failing, I don't see logs.

Thalley replied:
@giansta I think the server pinged out or something (happened to one of my PRs as well). I restarted the test :)

giansta replied:
Thanks, it was resolved later :-)

giansta commented on subsys/bluetooth/host/hci_core.h:
I followed your hints using both VS Set Scan Request Reports command and Scan Request Received Event.
However in my opinion the sample app needs also the host callback, so I would put it in a 2nd PR.

giansta commented on subsys/bluetooth/host/adv.c:
Same comment as above

cvinayak commented:
@giansta Please add `CONFIG_BT_CTLR_VS_SCAN_REQ_RX=y` after the line mentioned below, so the feature is covered in CI builds:
https://github.com/zephyrproject-rtos/zephyr/blob/beb1a0acaf86f9037002270bc0ab033a803ba460/tests/bluetooth/init/prj_ctlr_4_0_dbg.conf#L30

giansta committed:
Bluetooth: Controller: Vendor Specific: support for Scan Request Reports

Add support for vendor specific Set Scan Request Reports command and
Scan Request Received Event, available for legacy advertising.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

cvinayak commented:
@giansta 

> I tried such a change because our application using nRF52832 implements a BLE central that already uses most of the available 64k SRAM.

Also, upstream now supports Link Time Optimizations.

```
CONFIG_BT_CTLR_OPTIMIZE_FOR_SPEED=n

CONFIG_SHARED_INTERRUPTS=y
CONFIG_ISR_TABLES_LOCAL_DECLARATION=y
CONFIG_LTO=y
```

These configurations should help reduce overall flash usage of your application.

cvinayak commented:
> > @giansta Please add `CONFIG_BT_CTLR_VS_SCAN_REQ_RX=y` after the line mentioned below, so the feature is covered in CI builds:
> > https://github.com/zephyrproject-rtos/zephyr/blob/beb1a0acaf86f9037002270bc0ab033a803ba460/tests/bluetooth/init/prj_ctlr_4_0_dbg.conf#L30
> > 
> > Also fix the said line to `CONFIG_BT_HCI_VS_EXT=y`.
> > Thank you for contributing the feature.
> > You can now send a new PR for Host changes which we can start reviewing.
> 
> I have created PR #68659 for host changes. But I was worried that it depends on this one and it's using the same changes in the controller. So, as I was expecting, it's also pushing the same changes done here. Is this a problem? Or nothing happens if pushed after this one?

it is ok to have the commit from this PR in the host PR, you can rebase the host PR when this PR is merged.

(btw we are in v3.6 release feature freeze period, expect the PRs to be not merged until the release is complete, but we can get the needed approvals during this period).

giansta commented:
> @giansta
> 
> > I tried such a change because our application using nRF52832 implements a BLE central that already uses most of the available 64k SRAM.
> 
> Also, upstream now supports Link Time Optimizations.
> 
> ```
> CONFIG_BT_CTLR_OPTIMIZE_FOR_SPEED=n
> 
> CONFIG_SHARED_INTERRUPTS=y
> CONFIG_ISR_TABLES_LOCAL_DECLARATION=y
> CONFIG_LTO=y
> ```
> 
> These configurations should help reduce overall flash usage of your application.

We were already using `CONFIG_BT_CTLR_OPTIMIZE_FOR_SPEED`.
I tried with your suggestions.
 * Using CONFIG_SHARED_INTERRUPTS i got slightly more  flash usage
 * Using the last two, I get a "`undefined reference to _exit'`" error when using PICOLIBC or many warnings with DTS symbols when using `MINIMAL_LIBC`. 

But in all cases I get no RAM reduction (only 0.1% in last case) and Flash reduction only with last step. Yet our main issue is RAM. I'll try again if I'll find time to look into the liker documentation. Yet I'm already happy that all works even rebasing with latest released Zephyr 3.6.
